### DISCUSSION TITLE
Evaluate using Profile-Guided Optimization (PGO) for the project

### DISCUSSION THREAD

**Discussion by zamazan4ik** [farm-fe/farm#1419] [Ideas]
Hi!

I found the project via the article on HackerNews, and was interested in it since Farm cares about performance. Recently I checked Profile-Guided Optimization (PGO) improvements on multiple projects. The results are available [here](https://github.com/zamazan4ik/awesome-pgo/). Since this, I think it will be a good idea to try to optimize Farm even further by applying PGO on it.

I can suggest the following things to do:

* We need to perform PGO benchmarks on Farm tools. If it shows improvements - add a note about possible improvements to the documentation. Providing an easier way (e.g. a build option) to build scripts with PGO can be useful for the end-users too.
* Also, it will be a good idea to integrate building with PGO into your CI pipelines. In this case, users will get already PGO optimized binaries from the website (if you are going to provide prebuilt binaries, of course)

For the Rust projects, I suggest trying to start with [cargo-pgo](https://github.com/Kobzol/cargo-pgo).

[Here](https://github.com/zamazan4ik/awesome-pgo/) you can find different materials about PGO: benchmarks in different software, examples of how PGO is already integrated with different projects, PGO support in multiple Rust compilers, and some PGO-related pieces of advice.

After PGO, I suggest evaluating the LLVM BOLT optimizer - it can give more aggressive optimizations even after PGO. However, starting with regular PGO will be easier to do.

I would be happy to answer all your questions about PGO!

P.S. Did you think about enabling Link-Time Optimization (LTO) for the project? It can also help achieve better performance and smaller binaries.

**Comment by wre232114** [comment]
We're glad to  try PGO, thanks for the suggestion. And we're glad to enable LTO, PR welcome!
